### DISCUSSION TITLE
Quarkus Dependency Injection

### DISCUSSION THREAD

**Discussion by giscus** [quarkusio/quarkus#22077] [Quarkus Blog/Website]
# Quarkus Dependency Injection

Quarkus: Supersonic Subatomic Java

https://quarkus.io/blog/quarkus-dependency-injection/

**Comment by udaygowdg** [comment]
Hi Martin,

I have a jar which is developed using Java and Dagger Dependency Injections instead of CDI. 
When I use this jar as a dependency in my Quarkus application, throwing Unsatisfied dependency exceptions.

Does Quarkus recognizes only CDI dependencies ?

**Comment by mkouba** [comment]
Hi @udaygowdg,

> Does Quarkus recognizes only CDI dependencies ?

By default, [Quarkus discovers beans](https://quarkus.io/guides/cdi-reference#bean_discovery) from dependencies that contain a `beans.xml` descriptor and/or dependencies that contain a Jandex index - `META-INF/jandex.idx`. However, you can [exclude a dependency from discovery](https://quarkus.io/guides/cdi-reference#quarkus-arc_quarkus.arc.exclude-dependency-artifacts-that-should-be-excluded-from-discovery) or even [exclude specific types](https://quarkus.io/guides/cdi-reference#quarkus-arc_quarkus.arc.exclude-types).

**Comment by witmat** [comment]
Hi Martin

How can I mock final methods? I'm trying to mock final methods but I've got java.lang.VerifyError: class ****_ClientProxy overrides final method *****.close()V
How I can mock final methods while cdi blocks me?

**Reply by mkouba** [reply]
In general, CDI beans that need to be proxyable should not declare `final` methods. It's true that Quarkus attempts to remove the `final` modifier where necessary but I'm not sure how does it play together with the mocking libraries. @geoand Any idea?

In any case, if possible then remove the `final` modifiers from the bean class/methods.

**Reply by geoand** [reply]
Mocking libraries should work on the beans we remove the `final` modifier.

**Reply by witmat** [reply]
I can't remove final, it's in the external library. So quarkus will remove finals in the future? Because now I can't mock public final AccessSecretVersionResponse accessSecretVersion(SecretVersionName name) from class com.google.cloud.secretmanager.v1.SecretManagerServiceClient

**Reply by mkouba** [reply]
> So quarkus will remove finals in the future?

Quarkus does remove the `final` modifier from methods but it's possible that the bean class is not loaded by the Quarkus class loader. We'll need more info: full stack trace, quarkus version, code snippets, ideally a simple reproducer app.
